tests/utils/test_feature_utils.py

```diff
@@ -21,3 +21,110 @@ def test_normalize_coordinates(self):
         expected_coordinates = np.array([[-0.1, -0.1], [0.05, -0.18], [0.1, 0.03]])
 
         np.testing.assert_allclose(normalized_coordinates, expected_coordinates)
+
+    def test_convert_to_homogenous_coordinates_on_valid_input(self):
+        """Tests conversion to homogenous coordinates on valid input."""
+
+        non_homogenous_coordinates = np.random.rand(10, 2)
+        computed = feature_utils.convert_to_homogenous_coordinates(non_homogenous_coordinates)
+
+        np.testing.assert_allclose(computed[:, :2], non_homogenous_coordinates)
+        np.testing.assert_allclose(computed[:, 2], 1)
+
+    def test_convert_to_homogenous_coordinates_on_empty_input(self):
+        """Tests conversion to homogenous coordinates on empty input."""
+
+        non_homogenous_coordinates = np.array([])
+        computed = feature_utils.convert_to_homogenous_coordinates(non_homogenous_coordinates)
+
+        self.assertEqual(computed.size, 0)
+
+    def test_convert_to_homogenous_coordinates_on_none_input(self):
+        """Tests conversion to homogenous coordinates on None input."""
+
+        non_homogenous_coordinates = None
+        computed = feature_utils.convert_to_homogenous_coordinates(non_homogenous_coordinates)
+
+        self.assertIsNone(computed)
+
+    def test_convert_to_homogenous_coordinates_incorrect_dim(self):
+        """Tests conversion to homogenous coordinates on input which is already
+        3 dimensional."""
+
+        # test on invalid input
+        non_homogenous_coordinates = np.random.rand(5, 3)
+
+        self.assertRaises(
+            TypeError,
+            feature_utils.convert_to_homogenous_coordinates,
+            non_homogenous_coordinates,
+        )
+
+    def test_convert_to_epipolar_lines_valid_input(self):
+        """Test conversion of valid 2D points to epipolar lines using the essential matrix."""
+
+        points = np.array(
+            [
+                [10.0, -5.0],
+                [3.5, 20.0],
+            ]
+        )  # 2d points in homogenous coordinates
+        essential_mat = EssentialMatrix(Rot3.RzRyRx(0, np.deg2rad(45), 0), Unit3(np.array([-5, 2, 0])))
+
+        computed = feature_utils.convert_to_epipolar_lines(points, essential_mat)
+
+        expected = np.array(
+            [
+                [-2.36351579, -5.90878948, 1.75364193],
```

Comment:
where do these numbers come from? just doing the paper on math, or from a real example?

Reply:
I did math on paper upto two decimal places, checked it, and copied over the computed values for more precision